src/aof.c

```diff
@@ -928,7 +1311,106 @@ int loadAppendOnlyFile(char *filename) {
     if (fakeClient) freeClient(fakeClient);
     fclose(fp);
-    stopLoading(ret == AOF_OK);
+    return ret;
+}
+
+/* Load the AOF files according the aofManifest pointed by am. */
+int loadAppendOnlyFiles(aofManifest *am) {
+    serverAssert(am != NULL);
+    int ret = C_OK;
+    long long start;
+    off_t base_aof_size = 0, incr_aof_size = 0, size;
+    sds aof_name;
+
+    /* If there is no information about BASE AOF and INCR AOF in aofManifest, then 
+     * there is only one possibility: the aof manifest file does not exist, we may 
+     * be starting from an old redis version. so we must fall back to the previous 
+     * loading mode.
+     */
+    if (am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0) {
+        struct redis_stat sb;
+        /* If the `server.aof_filename` file does not exist, we will directly return 
+         * AOF_NOT_EXIST, and redis will ignore this error. */
+        if (redis_stat(server.aof_filename, &sb) != 0) {
+            return AOF_NOT_EXIST;
+        }
+
+        /* If the server.aof_filename file exists, we manually construct a BASE 
+         * type aofInfo and add it to aofManifest. In this way, we can reuse the 
+         * following code to load this AOF file. */
+        aofInfo *ai = aofInfoCreate();
+        ai->file_name = sdsnew(server.aof_filename);
+        ai->file_seq = 1;
+        ai->file_type = AOF_FILE_TYPE_BASE;
+        am->base_aof_info = ai;
+        am->curr_base_aof_seq = 1;
+        am->dirty = 1;
+
+        int ret = persistAofManifest(am);
+        if (ret == C_ERR) {
+            return AOF_FAILED;
+        }
+    }
+
+    /* Load BASE AOF. */
+    if (am->base_aof_info) {
+        serverAssert(am->base_aof_info->file_type == AOF_FILE_TYPE_BASE);
+        aof_name = (char*)am->base_aof_info->file_name;
+        size = getAppendOnlyFileSize(aof_name);
+        start = ustime();
+
+        startLoadingFile(size, aof_name, RDBFLAGS_AOF_PREAMBLE);
+        ret = loadSingleAppendOnlyFile(aof_name);
+        stopLoading(ret == AOF_OK);
+
+        if (ret == AOF_OK) {
+            base_aof_size = size;
+            serverLog(LL_NOTICE, "DB loaded from append only file %s: %.3f seconds",
+                aof_name, (float)(ustime()-start)/1000000);
+        }
+
+        /* If an AOF exists in the manifest but not on the disk, we consider 
+         * this to be a fatal error. */
+        if (ret == AOF_NOT_EXIST) ret = AOF_FAILED;
+
+        if (ret != AOF_OK && ret != AOF_EMPTY) {
+            return ret;
+        }
+    }
+
+    /* Load INCR AOFs. */
+    listNode *ln;
+    listIter li;
+
+    listRewind(am->incr_aof_list, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        aofInfo *ai = (aofInfo*)ln->value;
+        serverAssert(ai->file_type == AOF_FILE_TYPE_INCR);
+
+        aof_name = (char*)ai->file_name;
+        size = getAppendOnlyFileSize(aof_name);
+        start = ustime();
+        
+        startLoadingFile(size, aof_name, RDBFLAGS_AOF_PREAMBLE);
```

Comment:
Do you mean that all INCR AOF loads only use a pair of `startload/stopload`? Then use an function like `updateLoadSize` to update the file size in the middle?

Reply:
yes, just one pair of startLoad / stopLoad, ideally we should know the total size of all when we start loading, and fix the relevant places not to reset the read counters when we switch files.
thing is that the arguments for this startLoading function are used for two things:
1. the file name is used in order to know what to report on failures, and be able to distinguish between diskless and dis-based replication (see `rdbReportError`)
2. the size argument is used for progress reporting, see `INFO persistence`, i.e. `loading_total_bytes`, `loading_loaded_bytes`, `loading_loaded_perc`.

we need to somehow make them work correctly.

Reply:
Modified